libraries/state_history/include/eosio/state_history/log.hpp

```diff
@@ -62,57 +77,261 @@ struct state_history_log_header {
+   }
+
+   void construct_index(const fc::path& index_file_name) {
+      fc::cfile index_file;
+      index_file.set_file_path(index_file_name);
+      index_file.open("w+b");
+
+      uint64_t pos = 0;
+      while (pos < size()) {
+         uint64_t payload_size = payload_size_at(pos);
+         index_file.write(reinterpret_cast<const char*>(&pos), sizeof(pos));
+         pos += (sizeof(state_history_log_header) + payload_size + sizeof(uint64_t));
+      }
+   }
 };
 
+// directly adapt from boost/iostreams/filter/counter.hpp and change the type of chars_ to uint64_t.
+class counter  {
+public:
+    typedef char char_type;
+    struct category
+        : bio::dual_use,
+          bio::filter_tag,
+          bio::multichar_tag,
+          bio::optimally_buffered_tag
+        { };
+    explicit counter(uint64_t first_char = 0)
+        : chars_(first_char)
+        { }
+    uint64_t characters() const { return chars_; }
+    std::streamsize optimal_buffer_size() const { return 0; }
+
+    template<typename Source>
+    std::streamsize read(Source& src, char_type* s, std::streamsize n)
+    {
+        std::streamsize result = bio::read(src, s, n);
+        if (result == -1)
+            return -1;
+        chars_ += result;
+        return result;
+    }
+
+    template<typename Sink>
+    std::streamsize write(Sink& snk, const char_type* s, std::streamsize n)
+    {
+        std::streamsize result = bio::write(snk, s, n);
+        chars_ += result;
+        return result;
+    }
+private:
+    uint64_t chars_;
+};
+
+} // namespace detail
+
 class state_history_log {
  private:
    const char* const       name = "";
-   std::string             log_filename;
-   std::string             index_filename;
-   std::optional<state_history_log_prune_config> prune_config; //is set, log is in pruned mode
+   state_history_log_config config;
    fc::cfile               log;
    fc::cfile               index;
    uint32_t                _begin_block = 0;        //always tracks the first block available even after pruning
    uint32_t                _index_begin_block = 0;  //the first block of the file; even after pruning. it's what index 0 in the index file points to
    uint32_t                _end_block   = 0;
    chain::block_id_type    last_block_id;
 
-   std::thread                                                              thr;
-   std::atomic<bool>                                                        write_thread_has_exception = false;
-   std::exception_ptr                                                       eptr;
-   boost::asio::io_context                                                  ctx;
-   boost::asio::io_context::strand                                          work_strand{ctx};
-   boost::asio::executor_work_guard<boost::asio::io_context::executor_type> work_guard =
-       boost::asio::make_work_guard(ctx);
-   std::recursive_mutex                                                     mx;
+   mutable std::mutex      _mx;
```

Comment:
Strictly speaking, this is not a required change; however, it would be nice if we documented the member data and/or functions that NEEDED mutex protection.

Reply:
Thanks!

Reply:
Your welcome.